File: src/main/java/com/github/xumeng/java/problems/medium/problem0061/Solution0061.java
package com.github.xumeng.java.problems.medium.problem0061;

import com.github.xumeng.java.common.datastructure.ListNode;

public class Solution0061 {

    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) {
            return null;
        }
        int listLength = calculateListLength(head);
        k = k % listLength;
        while (k > 0) {
            head = moveLastToFirst(head);
            k--;
        }
        return head;
    }

    private int calculateListLength(ListNode head) {
        ListNode node = head;
        int length = 0;
        while (node != null) {
            length++;
            node = node.next;
        }
        return length;
    }

    private ListNode moveLastToFirst(ListNode head) {
        if (head == null) {
            return null;
        }
        if (head.next == null) {
            return head;
        }
        ListNode lastNode = null;
        ListNode node = head;
        while (node != null) {
            if (node.next != null && node.next.next == null) {
                lastNode = node.next;
                node.next = null;
            }
            node = node.next;
        }
        assert lastNode != null;
        lastNode.next = head;
        head = lastNode;
        return head;
    }

}

File: src/main/kotlin/com/github/xumeng/kotlin/problems/medium/problem0061/Solution0061.kt
package com.github.xumeng.kotlin.problems.medium.problem0061

import com.github.xumeng.kotlin.common.datastructure.ListNode

class Solution0061 {

    fun rotateRight(head: ListNode?, k: Int): ListNode? {
        if (head == null) {
            return null
        }
        val listLength = calculateListLength(head)
        var i = k % listLength
        var newHead: ListNode = head
        while (i > 0) {
            newHead = moveLastToFirst(newHead)
            i--
        }
        return newHead
    }

    private fun calculateListLength(head: ListNode?): Int {
        var node = head
        var length = 0
        while (node != null) {
            length++
            node = node.next
        }
        return length
    }

    private fun moveLastToFirst(head: ListNode): ListNode {
        if (head.next == null) {
            return head
        }
        var lastNode: ListNode? = null
        var node: ListNode? = head
        while (node != null) {
            val nextNode = node.next
            if (nextNode != null && nextNode.next == null) {
                lastNode = node.next
                node.next = null
            }
            node = node.next
        }
        if (lastNode == null) {
            return head
        }
        lastNode.next = head
        return lastNode
    }

}